src/modules/navigator/navigator_main.cpp

```diff
@@ -843,27 +843,44 @@ void Navigator::geofence_breach_check(bool &have_geofence_position_data)
 			snprintf(geofence_violation_warning, sizeof(geofence_violation_warning), "Geofence exceeded");
 		}
 
-		gf_violation_type.flags.dist_to_home_exceeded = !_geofence.isCloserThanMaxDistToHome(fence_violation_test_point(0),
+		gf_violation_type.dist_to_home_exceeded = !_geofence.isCloserThanMaxDistToHome(fence_violation_test_point(0),
 				fence_violation_test_point(1),
 				_global_pos.alt);
 
-		gf_violation_type.flags.max_altitude_exceeded = !_geofence.isBelowMaxAltitude(_global_pos.alt +
+		gf_violation_type.max_altitude_exceeded = !_geofence.isBelowMaxAltitude(_global_pos.alt +
 				vertical_test_point_distance);
 
-		gf_violation_type.flags.fence_violation = !_geofence.isInsidePolygonOrCircle(fence_violation_test_point(0),
-				fence_violation_test_point(1),
-				_global_pos.alt);
+		gf_violation_type.fence_violation = !_geofence.isInsidePolygonOrCircle(fence_violation_test_point(0),
+						    fence_violation_test_point(1),
+						    _global_pos.alt);
+		if(gf_violation_type.fence_violation){
+			gf_violation_type.buffer_violation = !_geofence.isInsideBufferZone(fence_violation_test_point(0),
```

Comment:
Hi @xinnad , your addition of the negation to this LOC is probably where we are thinking differently.  Your logic works as is, so does #20055.  Good to check this all out though!

Reply:
If you're happy with this, I'd like to do it this way since it follows the same logic as the original geofence. Thanks for looking through!

Reply:
That does clean things up. Nice find. I updated my PR with the simplified logic with the inner buffer.  It still doesn't fit within flash on a number of builds, so we'll be in a holding pattern until some flash is opened up.